lib/compat/wordpress-6.8/blocks.php

```diff
@@ -6,40 +6,85 @@
  */
 
 /**
- * Filters the block type arguments during registration to stabilize experimental block supports.
+ * Filters the block type arguments during registration to stabilize
+ * experimental block supports.
  *
- * This is a temporary compatibility shim as the approach in core is for this to be handled
- * within the WP_Block_Type class rather than requiring a filter.
+ * This is a temporary compatibility shim as the approach in core is for this
+ * to be handled within the WP_Block_Type class rather than requiring a filter.
  *
  * @param array $args Array of arguments for registering a block type.
  * @return array Array of arguments for registering a block type.
  */
 function gutenberg_stabilize_experimental_block_supports( $args ) {
-	if ( empty( $args['supports']['typography'] ) ) {
+	if ( empty( $args['supports'] ) ) {
 		return $args;
 	}
 
-	$experimental_typography_supports_to_stable = array(
-		'__experimentalFontFamily'     => 'fontFamily',
-		'__experimentalFontStyle'      => 'fontStyle',
-		'__experimentalFontWeight'     => 'fontWeight',
-		'__experimentalLetterSpacing'  => 'letterSpacing',
-		'__experimentalTextDecoration' => 'textDecoration',
-		'__experimentalTextTransform'  => 'textTransform',
+	$experimental_to_stable_keys = array(
+		'typography'           => array(
+			'__experimentalFontFamily'     => 'fontFamily',
+			'__experimentalFontStyle'      => 'fontStyle',
+			'__experimentalFontWeight'     => 'fontWeight',
+			'__experimentalLetterSpacing'  => 'letterSpacing',
+			'__experimentalTextDecoration' => 'textDecoration',
+			'__experimentalTextTransform'  => 'textTransform',
+		),
+		'__experimentalBorder' => 'border',
 	);
 
-	$current_typography_supports = $args['supports']['typography'];
-	$stable_typography_supports  = array();
+	$updated_supports = array();
+	foreach ( $args['supports'] as $support => $config ) {
+		// Add the support's config as is when it's not in need of stabilization.
+		if ( empty( $experimental_to_stable_keys[ $support ] ) ) {
+			$updated_supports[ $support ] = $config;
+			continue;
+		}
+
+		// Stabilize the support's key if needed e.g. __experimentalBorder => border.
+		if ( is_string( $experimental_to_stable_keys[ $support ] ) ) {
+			$stabilized_key = $experimental_to_stable_keys[ $support ];
+
+			// If there is no stabilized key present, use the experimental config as is.
+			if ( ! array_key_exists( $stabilized_key, $args['supports'] ) ) {
+				$updated_supports[ $stabilized_key ] = $config;
+				continue;
+			}
+
+			// Determine the order of keys, so the last defined can be preferred.
+			$key_positions      = array_flip( array_keys( $args['supports'] ) );
+			$experimental_index = $key_positions[ $support ] ?? -1;
+			$stabilized_index   = $key_positions[ $stabilized_key ] ?? -1;
+			$experimental_first = $experimental_index < $stabilized_index;
+
+			// Update support config, prefer the last defined value.
+			if ( is_array( $config ) ) {
```

Comment:
This is not at all a blocker for now, but in the JS implementation we check for an object, so there'd never be a match against the `align` support, where an array is a valid value. Here, there theoretically could be a match.

In practice it's not an issue because we don't have anything in `$experimental_to_stable_keys` that references supports where an array is a valid value, but just thought I'd mention it for the (very) unlikely event that at some distant point in the future we create an experimental support that does accept an array and then want to stabilize it further down the track. In that case, this condition might need to be updated.

For now, though, I think it's fine to leave as-is. And please excuse my verbose comment here 😄

Reply:
Appreciate you raising this one. It has vaguely been on my radar but I was struggling to come up with a real-word case where it should happen.

The `align` support is stable, so wouldn't be in the `$experimental_to_stable_keys` array. In that case, the guard clause at the beginning of the loop should just continue on, while using the `align` config as is.

In the end, rather than waste too much time on coming up with hypotheticals to test, I decided to leave this until it was a problem.

Does that allay your concern at all? What else am I missing?

Reply:
I meant to also add that in recent times, there's been a decided effort to not land block supports at all unless they were stable. To prevent being in the situation like we are with some typography and border supports. That's been the case for entire block support features like `shadow` or subfeatures like `textColumns` and `minHeight`.

Reply:
Absolutely, what you've described captures my thinking, too, so you're not missing anything here!

> In the end, rather than waste too much time on coming up with hypotheticals to test, I decided to leave this until it was a problem.

That's where I was coming from, too. I was mostly thinking out loud, which is to say, flagging something that could _theoretically_ be an issue at some distant point in the future. And it's at that distant point (if we're adding an experimental support that uses arrays) where I think it could be addressed, rather than now 🙂

Reply:
> If the approach to not landing experimental block supports continues, there shouldn't be a scenario with a new experimental block support that needs stabilizing.

Indeed — my vote is to avoid using the `__experimental` prefix for things in the future, too 👍

Reply:
👍 Having it on the record here will also flag it for our future selves or others, if it is really needed in the end.